### PR TITLE
Fixes #10551 - Allow users to search content view versions by version and repository

### PR DESCRIPTION
**Pull request by johnpmitsch**
Because version is stored in a major and minor column I had to write a custom function to search by version. All the operators that scoped search provides in auto complete are accounted for. Ran into issues using scoped search gem to search cvv by environment, Seems to be an problem with a has-many-through relationship.

### CODE REVIEW

**Comment by jlsherrill** [comment]
[test]

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 1]
```diff
@@ -17,17 +18,22 @@ class Api::V2::ContentViewVersionsController < Api::V2::ApiController
     param :version, String, :desc => N_("Filter versions by version number"), :required => false
     param :composite_version_id, :identifier, :desc => N_("Filter versions that are components in the specified composite version"), :required => false
     param_group :search, Api::V2::ApiController
+
     def index
+      versions = ContentViewVersion.readable  # need to get rid of this before PR
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb]: See your comment :)

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 2]
```diff
@@ -17,17 +18,22 @@ class Api::V2::ContentViewVersionsController < Api::V2::ApiController
...
+      includes = [:content_view, :environments, :composite_content_views, :history => :task]
+      sort = "#{versions.table_name}.major desc, #{versions.table_name}.minor desc"
+      respond(:collection => scoped_search(versions, sort, '', :includes => includes))
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb]: Assuming this is intended to be index_relation instead of versions?

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 3]
```diff
@@ -17,17 +18,22 @@ class Api::V2::ContentViewVersionsController < Api::V2::ApiController
     param :version, String, :desc => N_("Filter versions by version number"), :required => false
     param :composite_version_id, :identifier, :desc => N_("Filter versions that are components in the specified composite version"), :required => false
     param_group :search, Api::V2::ApiController
+
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb]: Can remove - we typically have the documentation not separated from the function definition

File: app/models/katello/content_view_version.rb [hunk 4]
```diff
@@ -41,7 +41,26 @@ class ContentViewVersion < Katello::Model
...
+          minor ||= 0
+          query = where("major #{operator} :major OR (major = :major AND minor #{operator} :minor)", :major => major, :minor => minor)
+        when minor.nil?
+          query = where("major #{operator} (:major)", :major => major)
+        else
+          query = where("major #{operator} (:major) and minor #{operator} (:minor)", :major => major, :minor => minor)
+        end
+        _, conditions = query.to_sql.split("WHERE")
+      end
+      { :conditions => conditions || ""}
```
Comment (ehelms) [inline app/models/katello/content_view_version.rb]: You could initialize conditions to empty string at the top of the function and simplify this declaration.

File: test/models/content_view_version_test.rb [hunk 5]
```diff
@@ -9,6 +9,7 @@ def setup
       @dev = create(:katello_environment,  :organization => @cvv.organization, :prior => @cvv.organization.library,     :name => 'dev')
       @beta = create(:katello_environment, :organization => @cvv.organization, :prior => @dev,                         :name => 'beta')
       @composite_version = ContentViewVersion.find(katello_content_view_versions(:composite_view_version_1))
+      @repo_version = ContentViewVersion.find(katello_content_view_versions(:library_view_version_1))
```
Comment (ehelms) [inline test/models/content_view_version_test.rb]: `repo_version` is a confusing variable name for me.

File: test/models/content_view_version_test.rb [hunk 6]
```diff
@@ -67,6 +68,32 @@ def test_docker_count
...
+    def test_search_equal_version
+      assert_includes ContentViewVersion.search_for("version = 1.0"), @cvv
+      assert_includes ContentViewVersion.search_for("version = 1"), @cvv
+    end
+
+    def test_search_compare_version
+      assert_includes ContentViewVersion.search_for("version >= 1.0"), @cvv
+      assert_includes ContentViewVersion.search_for("version <= 1.0"), @cvv
+      assert_includes ContentViewVersion.search_for("version < 1.1"), @cvv
+      assert_includes ContentViewVersion.search_for("version < 2"), @cvv
```
Comment (ehelms) [inline test/models/content_view_version_test.rb]: Do you think its worth testing the boundary case of `< 1` ?

Comment (ehelms) [inline test/models/content_view_version_test.rb]: What do you think about an assert that ensures the `>` and/or `<` operators return multiple results?

File: test/models/content_view_version_test.rb [hunk 7]
```diff
@@ -67,6 +68,32 @@ def test_docker_count
       assert_equal image_count, cvv.docker_image_count
       assert_equal tag_count, cvv.docker_tag_count
     end
+
+    def test_search_equal_version
+      assert_includes ContentViewVersion.search_for("version = 1.0"), @cvv
+      assert_includes ContentViewVersion.search_for("version = 1"), @cvv
```
Comment (ehelms) [inline test/models/content_view_version_test.rb]: What about an assert that ensures a search of just the major returns all the minors as well?

File: test/models/content_view_version_test.rb [hunk 8]
```diff
@@ -67,6 +69,36 @@ def test_docker_count
       assert_equal image_count, cvv.docker_image_count
       assert_equal tag_count, cvv.docker_tag_count
     end
+
+    def test_search_equal_version
+      assert_includes ContentViewVersion.search_for("version = 1.0"), @cvv
+      query = ContentViewVersion.search_for("version = 1")
+      assert [@cvv, @cvv_minor] & query == [@cvv, @cvv_minor]
```
Comment (ehelms) [inline test/models/content_view_version_test.rb]: This seems like you could achieve the same thing by just comparing the query to the array directly.

Comment (johnpmitsch) [inline test/models/content_view_version_test.rb]: It is not matching the array, I think it is returning more than cvv and cvv_minor, can I check if it returns more than 2 or more results?

File: app/controllers/katello/api/v2/content_view_versions_controller.rb [hunk 9]
```diff
@@ -18,16 +19,19 @@ class Api::V2::ContentViewVersionsController < Api::V2::ApiController
     param :composite_version_id, :identifier, :desc => N_("Filter versions that are components in the specified composite version"), :required => false
     param_group :search, Api::V2::ApiController
     def index
+      includes = [:content_view, :environments, :composite_content_views, :history => :task]
+      sort = "#{versions.table_name}.major desc, #{versions.table_name}.minor desc"
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_view_versions_controller.rb]: I think this should be `ContentViewVersion.table_name` ? Or, `resource_class.table_name`

**Comment by ehelms** [comment]
I noticed that when auto completing version, it only shows the major and non of the minor versions as options.

**Comment by ehelms** [comment]
ACK, merging, thanks @johnpmitsch !
